test/testpushnotifications.cpp

```diff
@@ -0,0 +1,351 @@
+
+        processTextMessageSpy.wait();
+        QCOMPARE(processTextMessageSpy.count(), 6);
+        socket = processTextMessageSpy.at(4).at(0).value<QWebSocket *>();
+        socket->sendTextMessage("err: Invalid credentials");
+
+        // Now the canNotAuthenticate Signal should be emitted
+        canNotAuthenticateSpy.wait();
+        QCOMPARE(canNotAuthenticateSpy.count(), 1);
+
+        // Now call setup again and it should do another connection attempt
+        pushNotifications->setup();
+        processTextMessageSpy.wait();
+        QCOMPARE(processTextMessageSpy.count(), 8);
+
+        auto userSent = processTextMessageSpy.at(6).at(1).toString();
+        auto passwordSent = processTextMessageSpy.at(7).at(1).toString();
+        QCOMPARE(userSent, user);
+        QCOMPARE(passwordSent, password);
+    }
+
+    void testAuthentication_correctCredentials_emitReady()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy processTextMessageSpy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(processTextMessageSpy.isValid());
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+        QSignalSpy readySpy(pushNotifications.get(), &OCC::PushNotifications::ready);
+        QVERIFY(readySpy.isValid());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication and send authenticated response
+        processTextMessageSpy.wait();
+        QCOMPARE(processTextMessageSpy.count(), 2);
+        const auto socket = processTextMessageSpy.at(0).at(0).value<QWebSocket *>();
+        socket->sendTextMessage("authenticated");
+
+        // Wait for ready signal
+        readySpy.wait();
+        QCOMPARE(readySpy.count(), 1);
+    }
+
+    void testOnWebSocketSslError_sslError_emitConnectionLost()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy processTextMessageSpy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(processTextMessageSpy.isValid());
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+        QSignalSpy connectionLostSpy(pushNotifications.get(), &OCC::PushNotifications::connectionLost);
+        QVERIFY(connectionLostSpy.isValid());
+
+        // Setup push notifications
+        pushNotifications->setup();
+        processTextMessageSpy.wait();
+
+        // FIXME: This a little bit ugly but I had no better idea how to trigger a error on the websocket client.
```

Comment:
Indeed that case will be very hard to test and you're in fact not *really* testing it. You'd have to make a much more elaborated server to reach that case. I'd personally call it a day and just not have that test until that becomes really an area of recurring issues in the field.

Reply:
I agree that this tests are not the best but I feel better if I test it that then not at all. This way I can at least be condifent that the sslError() signal will be emitted if the qwebsocket gets the signal.

Reply:
This is false confidence though. It's so synthetic and coupled to the private parts of the class that you're not testing the real thing in any way. The only thing you're testing is a single connect statement, that's a very low value test.

If you really want to test ssl errors for real that should be doable by enabling ssl on the fake server and then settomg a missing or invalide certificate on it.

Reply:
And don't take me wrong: I'm all for having more tests rather than less, still those tests need to be meaningful and not just anchoring on implementation details. In my experience such low value tests tend to either break when they shouldn't (false positives) or not break when they should (false negatives which is even worse). I wrote my fair share of those. :-)

Reply:
Ok:) should I still remove it now where we also need to make sure that isReady() returns false after a SSL error?

Reply:
Since you seem to really want to keep it, here is my proposal:
 1. remove the friend class on PushNotifications
 2. use findChildren on the PushNotifications object to get to the websocket.

https://doc.qt.io/qt-5/qobject.html#findChildren